### PR TITLE
Adding sync support

### PR DESCRIPTION
**Pull request by prklm10**
- Adding sync support
- Adding timeout of 10 min

### CODE REVIEW

**Review by shantanuk-browserstack — APPROVED** [review]

File: Percy/PercyOnAutomate.cs [hunk 1]
```diff
@@ -50,17 +52,21 @@ public void Screenshot(String name, IEnumerable<KeyValuePair<string, object>>? o
                       userOptions["consider_region_elements"] = elementIds;
                   }
               }
+              if(!userOptions.ContainsKey(sync)) {
+                userOptions["sync"] = false;
```
Comment (chinmay-browserstack) [inline Percy/PercyOnAutomate.cs:56]: Remove default value

File: Percy/lib/CliWrapper.cs [hunk 2]
```diff
@@ -100,15 +107,16 @@ internal static String PostScreenshot(string name, JObject tag, List<Tile> tiles
...
+          sync = sync
         };
         dynamic res = Request("/percy/comparison", JObject.FromObject(screenshotOptions));
         dynamic data = DeserializeJson<dynamic>(res.content);
         if (data.success.ToString().ToLower() != "true")
         {
           throw new Exception(data.error.ToString());
         }
-        return data.link.ToString();
+        return data;
```
Comment (chinmay-browserstack) [inline Percy/lib/CliWrapper.cs:119]: it should be data.data

File: Percy/lib/CliWrapper.cs [hunk 3]
```diff
@@ -162,11 +170,13 @@ internal static void PostPOAScreenshot(string name, string sessionId, string com
         {
           throw new Exception(data.error.ToString());
         }
+        return JObject.FromObject(data);
```
Comment (chinmay-browserstack) [inline Percy/lib/CliWrapper.cs:173]: timeout increase?

File: Percy/providers/AppAutomate.cs [hunk 4]
```diff
@@ -125,9 +132,8 @@ public override String Screenshot(String name, ScreenshotOptions options, String
       }
       finally
       {
-        ExecutePercyScreenshotEnd(name, percyScreenshotUrl, error);
+        ExecutePercyScreenshotEnd(name, percyScreenshotUrl, options.Sync, error);
```
Comment (chinmay-browserstack) [inline Percy/providers/AppAutomate.cs:135]: Why passing in end call?

File: Percy/lib/CliWrapper.cs (hunk 3 above)
Comment (prklm10) [reply Percy/lib/CliWrapper.cs:173]: it is in request function.

File: Percy/providers/AppAutomate.cs (hunk 4 above)
Comment (prklm10) [reply Percy/providers/AppAutomate.cs:135]: added this for instrumentation in seleniumHub. There is a selenium hub pr going there.

File: Percy/lib/CliWrapper.cs (hunk 2 above)
Comment (prklm10) [reply Percy/lib/CliWrapper.cs:119]: it is done in function that calls this.

File: Percy/PercyOnAutomate.cs (hunk 1 above)
Comment (prklm10) [reply Percy/PercyOnAutomate.cs:56]: removed

File: Percy/AppPercy.cs [hunk 5]
```diff
@@ -39,17 +40,22 @@ public void Screenshot(String name, ScreenshotOptions? options = null, Boolean f
...
       {
         GenericProvider provider;
         provider = ProviderResolver.ResolveProvider(percyAppiumDriver);
-        provider.Screenshot(
+        JObject data =  provider.Screenshot(
           name,
           options
         );
+
+        if (data?.TryGetValue("data", out JToken results) == true) {
```
Comment (ninadbstack) [inline Percy/AppPercy.cs:55]: do we have double `data` key ?  i see one extraction here and another in percyOnAutomate.cs line 56

File: Percy/IPercy.cs [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
 using System.Collections.Generic;
+using Newtonsoft.Json.Linq;
 
 namespace PercyIO.Appium
 {
   public interface IPercy
   {
-    void Screenshot(String name, ScreenshotOptions? options, Boolean fullScreen);
-    void Screenshot(String name, IEnumerable<KeyValuePair<string, object>>? options);
+    JObject Screenshot(String name, ScreenshotOptions? options, Boolean fullScreen);
```
Comment (ninadbstack) [inline Percy/IPercy.cs:9]: shouldnt this be ? [ makes it optional ], without `?` you should never be returning null

File: Percy/PercyOnAutomate.cs [hunk 7]
```diff
@@ -52,15 +52,17 @@ public void Screenshot(String name, IEnumerable<KeyValuePair<string, object>>? o
...
       }
       catch(Exception error)
       {
           Utils.Log($"Could not take Percy Screenshot \"{name}\"");
           Utils.Log(error.ToString(), "debug");
+          return null;
       }
     }
-    public void Screenshot(String name, ScreenshotOptions? options, bool fullScreen) {
+    public JObject Screenshot(String name, ScreenshotOptions? options, bool fullScreen) {
```
Comment (ninadbstack) [inline Percy/PercyOnAutomate.cs:65]: this is pretty confusing for users, need to consolidate. Can you raise with product team once ?

File: Percy/lib/CliWrapper.cs [hunk 8]
```diff
@@ -15,16 +15,17 @@ namespace PercyIO.Appium
...
     public static readonly string CLI_API = Environment.GetEnvironmentVariable("PERCY_CLI_API") ?? "http://localhost:5338";
-    private static HttpClient _http = new HttpClient();
+    private static HttpClient? _http;
     private static bool? _enabled = null;
 
     private static dynamic Request(string endpoint, JObject? payload = null)
     {
       StringContent? body = payload == null ? null : new StringContent(
                 payload.ToString(), Encoding.UTF8, "application/json");
+      HttpClient httpClient = getHttpClient();
```
Comment (ninadbstack) [inline Percy/lib/CliWrapper.cs:25]: [Optional] This is not the dotnet way in general, it should use Property

File: Percy/lib/CliWrapper.cs [hunk 9]
```diff
@@ -87,7 +88,7 @@ private static dynamic Request(string endpoint, JObject? payload = null)
       }
     };
 
-    internal static String PostScreenshot(string name, JObject tag, List<Tile> tiles, String externalDebugUrl, JObject ignoredElementsData, JObject consideredElementsData)
+    internal static dynamic PostScreenshot(string name, JObject tag, List<Tile> tiles, String externalDebugUrl, JObject ignoredElementsData, JObject consideredElementsData, Boolean? sync)
```
Comment (ninadbstack) [inline Percy/lib/CliWrapper.cs:91]: why is this dynamic while other is JObject ? shouldnt both be JObject ?

File: Percy/providers/GenericProvider.cs [hunk 10]
```diff
@@ -109,7 +109,8 @@ public virtual String Screenshot(String name, ScreenshotOptions options, String
         considerElementsData = consideredRegions
       });
       var tiles = CaptureTiles(options);
-      return CliWrapper.PostScreenshot(name, tag, tiles, debugUrl, ignoredElementsData, consideredElementsData);
+      var sync = options.Sync;
```
Comment (ninadbstack) [inline Percy/providers/GenericProvider.cs:112]: any specific reason here? all it does it sends it to next function

File: Percy/AppPercy.cs (hunk 5 above)
Comment (chinmay-browserstack) [reply Percy/AppPercy.cs:55]: This is for app percy flow and other which you have seen is for PoA flow

Comment (ninadbstack) [reply Percy/AppPercy.cs:55]: Oh, they have different flows. nevermind then!

Comment (ninadbstack) [reply Percy/AppPercy.cs:55]: Both are using different way of doing same thing so.

**Review by ninadbstack — APPROVED** [review]

**Pull request merged**